lib/python/qmk/flashers.py

```diff
@@ -0,0 +1,192 @@
+    port = _find_serial_port(details[0], details[1])
+    if port:
+        cli.run(['avrdude', '-p', 'atmega32u4', '-c', 'avr109', '-U', f'flash:w:{file}:i', '-P', port], capture_output=False)
+        return False
+    else:
+        return True
+
+
+def _flash_atmel_dfu(mcu, file):
+    force = '--force' if _check_dfu_programmer_version() else ''
+    cli.run(['dfu-programmer', mcu, 'erase', force], capture_output=False)
+    cli.run(['dfu-programmer', mcu, 'flash', force, file], capture_output=False)
+    cli.run(['dfu-programmer', mcu, 'reset'], capture_output=False)
+
+
+def _flash_hid_bootloader(mcu, details, file):
+    if details == 'halfkay':
+        if shutil.which('teensy-loader-cli'):
+            cmd = 'teensy-loader-cli'
+        elif shutil.which('teensy_loader_cli'):
+            cmd = 'teensy_loader_cli'
+
+    # Use 'hid_bootloader_cli' for QMK HID and as a fallback for HalfKay
+    if not cmd:
+        if shutil.which('hid_bootloader_cli'):
+            cmd = 'hid_bootloader_cli'
+        else:
+            return True
+
+    cli.run([cmd, f'-mmcu={mcu}', '-w', '-v', file], capture_output=False)
+
+
+def _flash_stm32(details, file):
+    # STM32duino
+    if details[0] == '1eaf' and details[1] == '0003':
+        cli.run(['dfu-util', '-a', '2', '-d', f'{details[0]}:{details[1]}', '-R', '-D', file], capture_output=False)
+    # STM32 DFU or APM32 DFU
+    else:
+        cli.run(['dfu-util', '-a', '0', '-d', f'{details[0]}:{details[1]}', '-s', '0x08000000:leave', '-D', file], capture_output=False)
+
+
+def _flash_isp(mcu, programmer, file):
+    programmer = 'usbasp' if programmer == 'usbasploader' else 'usbtiny'
+    # Check if the provide mcu has an avrdude-specific name, otherwise pass on what the user provided
+    mcu = AVRDUDE_MCU.get(mcu, mcu)
+    cli.run(['avrdude', '-p', mcu, '-c', programmer, '-U', f'flash:w:{file}:i'], capture_output=False)
+
+
+def flasher(mcu, file):
+    bl, details = _find_bootloader()
+    # Add a small sleep to avoid race conditions
+    time.sleep(1)
+    if bl == 'atmel-dfu':
+        _flash_atmel_dfu(details, file.name)
+    elif bl == 'caterina':
+        if _flash_caterina(details, file.name):
+            return (True, "The Caterina bootloader was found but is not writable. Check 'qmk doctor' output for advice.")
+    elif bl == 'hid-bootloader':
+        if mcu:
+            if _flash_hid_bootloader(mcu, details, file.name):
+                return (True, "Please make sure 'teensy_loader_cli' or 'hid_bootloader_cli' is available on your system.")
+        else:
+            return (True, "Specifying the MCU with '-m' is necessary for HalfKay/HID bootloaders!")
+    elif bl == 'stm32-dfu' or bl == 'apm32-dfu':
+        _flash_stm32(details, file.name)
+    elif bl == 'usbasploader' or bl == 'usbtinyisp':
+        if mcu:
+            _flash_isp(mcu, bl, file.name)
+        else:
+            return (True, "Specifying the MCU with '-m' is necessary for ISP flashing!")
+
```

Comment:
Looks like the `kiibohd` bootloader type is not handled here, even though `BOOTLOADER_VIDS_PIDS` has an entry for it. I don't have the corresponding hardware, and apparently there is some difference between what `qmk flash` would use for that bootloader:
```makefile
    DFU_ARGS = -d 1C11:B007
```
and [what QMK Toolbox would use](https://github.com/qmk/qmk_toolbox/blob/ad6d19579be8b1a0e0713080af857914886c674e/windows/QMK%20Toolbox/Usb/Bootloader/KiibohdDfuDevice.cs#L20):
```
await RunProcessAsync("dfu-util.exe", $"-a 0 -d 1C11:B007 -D \"{file}\"");
```
(here the extra `-a 0` option is added).

Another bootloader type that could potentially be supported here is `gd32v-dfu`, but the IDs for it are not present in both `BOOTLOADER_VIDS_PIDS` and udev rules.

`wb32-dfu` could probably wait for now (it needs special tools that are probably not packaged in most distros, and `qmk_firmware` apparently does not even have USB IDs for it, although [QMK Toolbox apparently has](https://github.com/qmk/qmk_toolbox/commit/7b113f596f52412bfb316415225c68226522f3e0#diff-53f54b93f00bd69b0f4f1ec9b02a5f96f82c0c9a6815a1b3f2200e8238fa72aeR308-R313)).

Reply:
> (here the extra -a 0 option is added).

`_flash_stm32` already adds `-a 0` anyway, not sure we need anything further?

Reply:
replicated the kiibohd command explicitly as per toolbox.

Reply:
> replicated the kiibohd command explicitly as per toolbox.

That would probably be the safest thing to do (in the future we could fix `builddefs/bootloader.mk` to use exactly the same options for this bootloader too).